src/egit-checkout.c

```diff
@@ -0,0 +1,243 @@
+#include <string.h>
+
+#include "git2.h"
+
+#include "egit.h"
+#include "interface.h"
+#include "egit-checkout.h"
+
+
+// =============================================================================
+// Helpers - Checkout Options
+
+typedef struct {
+    emacs_env *env;
+    emacs_value callback;
+} checkout_options_ctx;
+
+static int checkout_notify_callback(
+    git_checkout_notify_t why,
+    const char *path,
+    __attribute__((unused)) const git_diff_file *baseline,
+    __attribute__((unused)) const git_diff_file *target,
+    __attribute__((unused)) const git_diff_file *workdir,
+    void *payload)
+{
+    checkout_options_ctx *ctx = (checkout_options_ctx*) payload;
+    emacs_env *env = ctx->env;
+
+    emacs_value args[2];
+    args[0] = em_nil;
+    args[1] = EM_STRING(path);
+
+    switch (why) {
+    case GIT_CHECKOUT_NOTIFY_IGNORED:
+        args[0] = em_ignored;
+        break;
+    case GIT_CHECKOUT_NOTIFY_UNTRACKED:
+        args[0] = em_untracked;
+        break;
+    case GIT_CHECKOUT_NOTIFY_UPDATED:
+        args[0] = em_updated;
+        break;
+    case GIT_CHECKOUT_NOTIFY_DIRTY:
+        args[0] = em_dirty;
+        break;
+    case GIT_CHECKOUT_NOTIFY_CONFLICT:
+        args[0] = em_conflict;
+        break;
+    default: break;
+    }
+
+    emacs_value retval = env->funcall(env, ctx->callback, 2, args);
+    EM_RETURN_IF_NLE(GIT_EUSER);
+
+    if (EM_EQ(retval, em_abort))
+        return GIT_EUSER;
+    return 0;
+}
+
+static void checkout_progress_callback(
+    const char *path,
+    size_t completed_steps,
+    size_t total_steps,
+    void *payload)
+{
+    checkout_options_ctx *ctx = (checkout_options_ctx*) payload;
+    emacs_env *env = ctx->env;
+
+    emacs_value args[3];
+    args[0] = EM_STRING(path);
+    args[1] = EM_INTEGER(completed_steps);
+    args[2] = EM_INTEGER(total_steps);
+
+    env->funcall(env, ctx->callback, 3, args);
+
+    // TODO: Since we can't abort from inside a progress callback, we
```

Comment:
What exactly is the TODO here? What action remains to be taken?

Reply:
I just need to remind myself to confirm that it works the way I think it does, write a test for it and document it, basically.